src/streaming/utils/BaseURLSelector.js

```diff
@@ -0,0 +1,131 @@
+ * included below. This software may be subject to other third party and contributor
+ * rights, including patent rights, and no such rights are granted under this license.
+ *
+ * Copyright (c) 2013, Dash Industry Forum.
+ * All rights reserved.
+ *
+ * Redistribution and use in source and binary forms, with or without modification,
+ * are permitted provided that the following conditions are met:
+ *  * Redistributions of source code must retain the above copyright notice, this
+ *  list of conditions and the following disclaimer.
+ *  * Redistributions in binary form must reproduce the above copyright notice,
+ *  this list of conditions and the following disclaimer in the documentation and/or
+ *  other materials provided with the distribution.
+ *  * Neither the name of Dash Industry Forum nor the names of its
+ *  contributors may be used to endorse or promote products derived from this software
+ *  without specific prior written permission.
+ *
+ *  THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS AS IS AND ANY
+ *  EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE IMPLIED
+ *  WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE ARE DISCLAIMED.
+ *  IN NO EVENT SHALL THE COPYRIGHT HOLDER OR CONTRIBUTORS BE LIABLE FOR ANY DIRECT,
+ *  INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING, BUT
+ *  NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES; LOSS OF USE, DATA, OR
+ *  PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY,
+ *  WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE)
+ *  ARISING IN ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE
+ *  POSSIBILITY OF SUCH DAMAGE.
+ */
+
+import EventBus from '../../core/EventBus.js';
+import Events from '../../core/events/Events.js';
+import DashManifestModel from '../../dash/models/DashManifestModel.js';
+import BlacklistController from '../controllers/BlacklistController.js';
+import DVBSelector from '../rules/baseUrlResolution/DVBSelector.js';
+import BasicSelector from '../rules/baseUrlResolution/BasicSelector.js';
+import FactoryMaker from '../../core/FactoryMaker.js';
+
+const URL_RESOLUTION_FAILED_GENERIC_ERROR_CODE = 1;
+const URL_RESOLUTION_FAILED_GENERIC_ERROR_MESSAGE = 'Failed to resolve a valid URL';
+
+function BaseURLSelector() {
+
+    const context = this.context;
+    const eventBus = EventBus(context).getInstance();
+    const dashManifestModel = DashManifestModel(context).getInstance();
+
+    let instance,
+        serviceLocationBlacklistController,
+        basicSelector,
+        dvbSelector,
+        selector;
+
+    function setup() {
+        serviceLocationBlacklistController = BlacklistController(context).create({
+            updateEventName:        Events.SERVICE_LOCATION_BLACKLIST_CHANGED,
+            loadFailedEventName:    Events.FRAGMENT_LOADING_COMPLETED
```

Comment:
While this reads strange, that does make sense... it would be could if the download error that is sent out to the client is annotated with whether more options are available. Or exposing URL_RESOLUTION_FAILED would likely achieve give the client the information it would need to make a decision on whether to do something.